src/d3a/models/strategy/external_strategies/storage.py

```diff
@@ -0,0 +1,261 @@
+                                             self.connected):
+            return
+        try:
+            arguments = json.loads(payload["data"])
+            assert set(arguments.keys()) == {'offer'}
+        except Exception as e:
+            logging.error(f"Error when handling delete offer request. Payload {payload}. "
+                          f"Exception {str(e)}.")
+            self.redis.publish_json(
+                delete_offer_response_channel,
+                {"error": "Incorrect delete offer request. Available parameters: (offer)."}
+            )
+        else:
+            self.pending_requests.append(
+                IncomingRequest("delete_offer", arguments, delete_offer_response_channel))
+
+    def _delete_offer_impl(self, arguments, response_channel):
+        try:
+            self.market.delete_offer(arguments["offer"])
+            self.offers.remove_by_id(arguments["offer"])
+            self.redis.publish_json(response_channel,
+                                    {"status": "ready", "deleted_offer": arguments["offer"]})
+        except Exception as e:
+            logging.error(f"Error when handling offer delete on area {self.device.name}: "
+                          f"Exception: {str(e)}, Offer Arguments: {arguments}")
+            self.redis.publish_json(
+                response_channel,
+                {"status": "error",
+                 "error_message": f"Error when handling offer delete "
+                                  f"on area {self.device.name} with arguments {arguments}."})
+
+    def _offer(self, payload):
+        offer_response_channel = f'{self.device.name}/offer/response'
+        if not check_for_connected_and_reply(self.redis, offer_response_channel,
+                                             self.connected):
+            return
+        try:
+            arguments = json.loads(payload["data"])
+            assert set(arguments.keys()) == {'price', 'energy'}
+            arguments['seller'] = self.device.name
+            arguments['seller_origin'] = self.device.name
+        except Exception as e:
+            logging.error(f"Incorrect offer request. Payload {payload}. Exception {str(e)}.")
+            self.redis.publish_json(
+                offer_response_channel,
+                {"error": "Incorrect offer request. Available parameters: (price, energy)."}
+            )
+        else:
+            self.pending_requests.append(
+                IncomingRequest("offer", arguments, offer_response_channel))
+
+    def _offer_impl(self, arguments, response_channel):
+        try:
+            energy_sell_dict = self.state.clamp_energy_to_sell_kWh([self.market.time_slot])
+            assert energy_sell_dict[self.market.time_slot] >= arguments['energy']
+            expected_energy = self.state.used_storage * (1 - self.state.min_allowed_soc_ratio)
+            assert arguments['energy'] <= expected_energy
```

Comment:
Same as below, these assertions should be moved in the _offer method.

Reply:
Okay, then I assume, I also had to delete the offer that is posted [here](https://github.com/gridsingularity/d3a/pull/639/files#diff-1d553c99bd95475906818e1177a93f29R114) too.

Reply:
I think you misunderstood. There is no reason to delete any offer, since the offer is not posted yet. As mentioned below, what I suggest is to move this assert to the _offer method, which does not place an offer to the market, but only creates an incoming request that invokes _offer_impl on the event_tick.